saml2/saml2-service-provider/src/opensaml4Main/java/org/springframework/security/saml2/provider/service/authentication/OpenSaml4AuthenticationProvider.java

```diff
@@ -737,7 +801,14 @@ protected ValidationResult validateAddress(SubjectConfirmation confirmation, Ass
 				@Override
 				protected ValidationResult validateInResponseTo(SubjectConfirmation confirmation, Assertion assertion,
 						ValidationContext context, boolean required) {
-					// applications should validate their own in response to
+					String inResponseTo = confirmation.getSubjectConfirmationData().getInResponseTo();
```

Comment:
Very good idea, didn't realize how this class worked until now. This will of course require us to ALWAYS parse the request twice since we won't know beforehand if there are `InResponseTo` properties in the assertions or not. You think that's ok?

Reply:
I don't think that's quite true. I think you would only add the `SC_VALID_IN_RESPONSE_TO` parameter in the event that the assertion has a subject confirmation data has an `InResponseTo` attribute. For example, you can change the if statement on line 756 to check for the `InResponseTo` field. If it passes, parse the `AuthnRequest` and supply the id.